internal/controller/operator/factory/vmdistributedcluster/vmdistributedcluster.go

```diff
@@ -0,0 +1,563 @@
+		}
+	}
+
+	if !needsUpdate {
+		return nil
+	}
+
+	// Only update if the new list of target refs is actually different from the old one
+	// This helps prevent unnecessary updates if the needsUpdate logic had a nuance.
+	if reflect.DeepEqual(freshVMUserObj.Spec.TargetRefs, newTargetRefs) {
+		return nil
+	}
+
+	freshVMUserObj.Spec.TargetRefs = newTargetRefs
+	if err := rclient.Update(ctx, freshVMUserObj); err != nil {
+		return fmt.Errorf("failed to update vmuser %s: %w", freshVMUserObj.Name, err)
+	}
+
+	return nil
+}
+
+func waitForVMClusterReady(ctx context.Context, rclient client.Client, vmCluster *vmv1beta1.VMCluster, deadline time.Duration) error {
+	var lastStatus vmv1beta1.UpdateStatus
+	// Fetch VMCluster in a loop until it has UpdateStatusOperational status
+	err := wait.PollUntilContextTimeout(ctx, time.Second, deadline, true, func(ctx context.Context) (done bool, err error) {
+		if err := rclient.Get(ctx, types.NamespacedName{Name: vmCluster.Name, Namespace: vmCluster.Namespace}, vmCluster); err != nil {
+			if k8serrors.IsNotFound(err) {
+				return false, fmt.Errorf("VMCluster not found")
+			}
+			return false, fmt.Errorf("failed to fetch VMCluster %s/%s: %w", vmCluster.Namespace, vmCluster.Name, err)
+		}
+		lastStatus = vmCluster.Status.UpdateStatus
+		return vmCluster.Status.UpdateStatus == vmv1beta1.UpdateStatusOperational, nil
+	})
+	if err != nil {
+		return fmt.Errorf("failed to wait for VMCluster %s/%s to be ready: %w, current status: %s", vmCluster.Namespace, vmCluster.Name, err, lastStatus)
+	}
+
+	return nil
+}
+
+// VMAgentMetrics defines the interface for VMAgent objects that can provide metrics URLs
+type VMAgentMetrics interface {
+	AsURL() string
+	GetMetricPath() string
+}
+
+// VMAgentWithStatus extends VMAgentMetrics to include status checking
+type VMAgentWithStatus interface {
+	VMAgentMetrics
+	GetReplicas() int32
+	GetNamespace() string
+	GetName() string
+}
+
+// vmAgentAdapter wraps VMAgent to implement VMAgentWithStatus interface
+type vmAgentAdapter struct {
+	*vmv1beta1.VMAgent
+}
+
+func (v *vmAgentAdapter) GetReplicas() int32 {
+	return v.Status.Replicas
+}
+
+// GetName and GetNamespace are inherited from vmv1beta1.VMAgent
+
+func waitForVMClusterVMAgentMetrics(ctx context.Context, httpClient *http.Client, vmAgent VMAgentWithStatus, deadline time.Duration) error {
+	if vmAgent == nil {
+		// Don't throw error if VMAgent is nil, just exit early
+		return nil
+	}
+
+	if vmAgent.GetReplicas() == 0 {
+		return fmt.Errorf("VMAgent %s/%s is not ready", vmAgent.GetNamespace(), vmAgent.GetName())
+	}
+
+	parts := []string{vmAgent.AsURL(), vmAgent.GetMetricPath()}
+	vmAgentPath := strings.Join(parts, "")
+
+	// Loop until VMAgent metrics show that disk buffer is empty
+	err := wait.PollUntilContextTimeout(ctx, 1*time.Second, deadline, true, func(ctx context.Context) (done bool, err error) {
+		metricValue, err := fetchVMAgentDiskBufferMetric(ctx, httpClient, vmAgentPath)
```

Comment:
The vmAgentPath uses vmagent service address, what if the vmagent deployment has multiple replicas?